### DISCUSSION TITLE
Oscillations in INS velocity field

### DISCUSSION THREAD

**Discussion by ABallisat** [idaholab/moose#20499] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

I am using the FE implementation of the Navier Stokes module with two phase flows, using the Level Set method for interface tracking. I am having some issues with oscillations in the vertical component of the velocity field, see image below. The scenario for this test is just steady state with a less dense fluid sat above a denser fluid. All boundaries except the top are no-penetration (so v dot n = 0 on the boundaries of the mesh) and the pressure on the top boundary is fixed to zero (it also displays the same behaviour without applying it). I am using first order elements for all variables and running with PSPG and SUPG. This scenario is just testing that it produces a stable result, I would expect the velocities to tend to zero but instead these variations are observed and over time their magnitude increases until it will no longer solve.

Am I missing something obvious?

Thanks,

Alex

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/40231544/157288160-451a2717-27db-4ace-abe0-58e1109cb806.png">

**Comment by GiudGiud** [comment, marked as answer]
@lindsayad who can we tag on level set?

**Reply by lindsayad** [reply]
@aeslaughter 😄 I don't think any of our current team members have done much development in that module

**Reply by lindsayad** [reply]
@ABallisat how do things look if you make the two phases the same? (densities, etc.)

**Reply by ABallisat** [reply]
Same densities work to what I take is numerical error:

<img width="1535" alt="image" src="https://user-images.githubusercontent.com/40231544/157419683-c3b77310-a2fc-4a4e-b023-8c2ecf1a3c63.png">

In the original post, the densities ratio was O(1:1000), I also tried O(1:4) (not shown) which also displays similar oscillatory behaviour so I guess it comes down to the densities. Based on that, is the FE INS implementation designed to work with variable density?

**Reply by lindsayad** [reply]
Both convection and time derivative terms (in MOOSE's INS kernels in `navier_stokes`) assume that the density can be pulled outside the differential operator (d/dt and div respectively) which reflects an assumption of constant density.

Are you creating your own level set kernels, or are you using the `level_set` module?

**Reply by ABallisat** [reply]
I've had a good dig through the literature and there is a mix of formulations, some authors claim to use pulling out the density of the differential operators whilst others include them. At the moment I have not written kernels with them included in the differential operators but I think it would be sensible to see if this has an effect so I will attempt to implement this.

As for level set, I have been using the `level_set` module for propagation but given the reinitialisation implementation seems to not work at times, I have implemented a parallel fast marching implementation to do that bit.

**Reply by GiudGiud** [reply]
@ABallisat any update on this?

**Reply by ABallisat** [reply]
Apologies for just getting around to replying, I have been busy on other things recently. Long story short no, I had a good go at it but could never get it to converge. For now I have switched my attention to the finite volume implementation and seeing if that is any more suited to it. I played around with finite volume a long while ago but hit the caching problem so the memory usage was too high and it did not scale up. I have seen that that has now been fixed so will be interesting to see how much that has improved. I will report back on how I get on.

**Reply by GiudGiud** [reply]
Sounds good. 
Closing this for now. 
We'd be interested in an implementation of phase field in finite volume if you can do the conversion.
Do we have a final conclusion on whether different densities are supported?

**Reply by lindsayad** [reply]
I said that the INS FE implementation assumes that density is constant such that you can pull it out of derivative terms, so any spatial or temporal derivatives related to density would have to be added by the user if they wanted to also use the INS FE kernels. 

That being said, density in INS FE is a material property, so you could in theory "use" the INS FE objects with a non-constant density. It's just that omission of the additional derivative terms related to density would lead to an incorrect representation of the NS physics
